core/debug.cc

```diff
@@ -139,125 +139,113 @@ static const char *si_code_to_str(int sig_num, int si_code) {
   return "si_code unavailable for unknown signal";
 }
 
-/* addr2line must be available.
- * prints out the code lines [lineno - context, lineno + context] */
-static std::string print_code(char *symbol, int context) {
+//
+static std::string FetchLine(std::string filename, int lineno, int context) {
   std::ostringstream ret;
-  char executable[1024];
-  char addr[1024];
+  std::string line;
 
-  char cmd[1024];
+  std::ifstream f(filename);
+  if (!f) {
+    return "        (file/line not available)\n";
+  }
 
-  FILE *proc;
+  for (int curr = 1; std::getline(f, line) && curr <= lineno + context;
+       curr++) {
+    if (std::abs(curr - lineno) <= context) {
+      ret << "      " << (curr == lineno ? "->" : "  ") << " " << curr << ": "
+          << line << std::endl;
+    }
+  }
 
-  /* Symbol examples:
-   * ./bessd(run_worker+0x8e) [0x419d0e]" */
-  /* ./bessd() [0x4149d8] */
-  sscanf(symbol, "%[^(](%*s [%[^]]]", executable, addr);
+  return ret.str();
+}
 
-  snprintf(cmd, sizeof(cmd), "addr2line -C -i -f -p -e %s %s 2> /dev/null",
-           executable, addr);
+// Run an external command and return its standard output.
+static std::stringstream RunCommand(std::string cmd) {
+  std::stringstream ret;
 
-  proc = popen(cmd, "r");
-  if (!proc) {
-    return "";
+  FILE *proc = popen(cmd.c_str(), "r");
+  if (proc) {
+    while (!feof(proc)) {
+      char buf[PIPE_BUF];
+      size_t bytes = fread(buf, 1, sizeof(buf), proc);
+      ret.write(buf, bytes);
+    }
+    pclose(proc);
   }
 
-  for (;;) {
-    char line[1024];
-    char *p;
+  return ret;
```

Comment:
`std::stringstream` is not copyable, and its move constructor is not implemented until GCC 5 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54316). Perhaps return a `string` instead?